# Task
Load the data file "My Drive/customer_data.csv" into a dataframe and display the first 5 rows and the columns and their types.

## Mount google drive

### Subtask:
Mount your Google Drive to access files stored there.


**Reasoning**:
Mount Google Drive to access the data file.



## Load data

### Subtask:
Load the data from the specified path in your Google Drive into a pandas DataFrame.


**Reasoning**:
Import pandas and load the CSV file into a DataFrame.



In [1]:
%pip install pandas


[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import pandas as pd

df = pd.read_csv("C:\\Users\\vinay\\OneDrive\\Desktop\\hackathon\\autopharma\\dataset\\merged_data.csv")

**Reasoning**:
Attempt to load the data using a try-except block to handle `FileNotFoundError`.



## Display data

### Subtask:
Display the head of the DataFrame to verify that the data is loaded correctly.


**Reasoning**:
Display the first 5 rows and the column names and their data types of the DataFrame to verify that the data is loaded correctly.



In [3]:
display(df.head())
display(df.info())

COSMIC_ID CELL_LINE_NAME TCGA_DESC  DRUG_ID     DRUG_NAME   LN_IC50  \
0     683667         PFSK-1        MB     1003  Camptothecin -1.463887   
1     687448       COLO-829      SKCM     1003  Camptothecin -1.235034   
2     687455            RT4      BLCA     1003  Camptothecin -2.963191   
3     687457          SW780      BLCA     1003  Camptothecin -1.449138   
4     687459         TCCSUP      BLCA     1003  Camptothecin -2.350633   

        AUC   Z_SCORE GDSC Tissue descriptor 1 GDSC Tissue descriptor 2  ...  \
0  0.930220  0.433123           nervous_system          medulloblastoma  ...   
1  0.867348  0.557727                     skin                 melanoma  ...   
2  0.821438 -0.383200        urogenital_system                  Bladder  ...   
3  0.905050  0.441154        urogenital_system                  Bladder  ...   
4  0.843430 -0.049682        urogenital_system                  Bladder  ...   

  Copy Number Alterations (CNA) Gene Expression_y Methylation_y  \
0                             Y                 Y             Y   
1                             Y                 Y             Y   
2                             Y                 Y             Y   
3                             Y                 Y             Y   
4                             Y                 Y             Y   

  Drug\nResponse GDSC\nTissue descriptor 1 GDSC\nTissue\ndescriptor 2  \
0              Y            nervous_system            medulloblastoma   
1              Y                      skin                   melanoma   
2              Y         urogenital_system                    Bladder   
3              Y         urogenital_system                    Bladder   
4              Y         urogenital_system                    Bladder   

  Cancer Type\n(matching TCGA label)  \
0                                 MB   
1                               SKCM   
2                               BLCA   
3                               BLCA   
4                               BLCA   

  Microsatellite \ninstability Status (MSI) Screen Medium_y  \
0                                 MSS/MSI-L               R   
1                                 MSS/MSI-L               R   
2                                 MSS/MSI-L           D/F12   
3                                 MSS/MSI-L           D/F12   
4                                 MSS/MSI-L           D/F12   

  Growth Properties_y  
0            Adherent  
1            Adherent  
2            Adherent  
3            Adherent  
4            Adherent  

[5 rows x 32 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166644 entries, 0 to 166643
Data columns (total 32 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   COSMIC_ID                                 166644 non-null  int64  
 1   CELL_LINE_NAME                            166644 non-null  object 
 2   TCGA_DESC                                 166644 non-null  object 
 3   DRUG_ID                                   166644 non-null  int64  
 4   DRUG_NAME                                 166644 non-null  object 
 5   LN_IC50                                   166644 non-null  float64
 6   AUC                                       166644 non-null  float64
 7   Z_SCORE                                   166644 non-null  float64
 8   GDSC Tissue descriptor 1                  166644 non-null  object 
 9   GDSC Tissue descriptor 2                  166644 non-null  object 
 10  Cancer Type (matchin

None

# Task
Perform feature engineering on the loaded dataset, including scaling numerical features, encoding categorical features, and optionally integrating drug features and applying dimensionality reduction.

## Identify numerical features

### Subtask:
Identify the continuous numerical features in the dataset.


**Reasoning**:
Identify numerical columns by checking the data types and filter for numerical types.



In [4]:
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
print("Numerical columns:")
print(numerical_cols)

Numerical columns:
['COSMIC_ID', 'DRUG_ID', 'LN_IC50', 'AUC', 'Z_SCORE', 'COSMIC identifier']


## Scale numerical features

### Subtask:
Normalize or scale the identified numerical features using a suitable method like standardization or min-max scaling.


**Reasoning**:
Scale the numerical features using StandardScaler and create a new DataFrame with the scaled features.



In [5]:
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [6]:
import sklearn
print(sklearn.__version__)


1.7.2


In [7]:
from sklearn.preprocessing import StandardScaler 

numerical_cols = ['COSMIC_ID', 'DRUG_ID', 'LN_IC50', 'AUC', 'Z_SCORE', 'COSMIC identifier']
scaler = StandardScaler()
df_scaled_numerical = pd.DataFrame(scaler.fit_transform(df[numerical_cols]), columns=numerical_cols)
display(df_scaled_numerical.head())

COSMIC_ID   DRUG_ID   LN_IC50       AUC   Z_SCORE  COSMIC identifier
0  -1.348967 -1.364272 -1.523914  0.333408  0.390658          -1.348967
1  -1.332118 -1.364272 -1.442914 -0.091805  0.515240          -1.332118
2  -1.332087 -1.364272 -2.054572 -0.402301 -0.425523          -1.332087
3  -1.332078 -1.364272 -1.518694  0.163179  0.398688          -1.332078
4  -1.332069 -1.364272 -1.837766 -0.253566 -0.092063          -1.332069

## Identify categorical features

### Subtask:
Identify the categorical features in the dataset.


**Reasoning**:
Select columns with 'object' data type, convert to a list, and print the list.



In [8]:
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()
print("Categorical columns:")
print(categorical_cols)

Categorical columns:
['CELL_LINE_NAME', 'TCGA_DESC', 'DRUG_NAME', 'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2', 'Cancer Type (matching TCGA label)', 'Microsatellite instability Status (MSI)', 'Screen Medium_x', 'Growth Properties_x', 'CNA', 'Gene Expression_x', 'Methylation_x', 'TARGET', 'TARGET_PATHWAY', 'Sample Name', 'Whole Exome Sequencing (WES)', 'Copy Number Alterations (CNA)', 'Gene Expression_y', 'Methylation_y', 'Drug\nResponse', 'GDSC\nTissue descriptor 1', 'GDSC\nTissue\ndescriptor 2', 'Cancer Type\n(matching TCGA label)', 'Microsatellite \ninstability Status (MSI)', 'Screen Medium_y', 'Growth Properties_y']


## Encode categorical features

### Subtask:
Apply one-hot encoding to the identified categorical variables.


**Reasoning**:
Apply one-hot encoding to the identified categorical columns and display the head of the resulting DataFrame.



In [9]:
categorical_cols = ['CELL_LINE_NAME', 'TCGA_DESC', 'DRUG_NAME', 'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2', 'Cancer Type (matching TCGA label)', 'Microsatellite instability Status (MSI)', 'Screen Medium_x', 'Growth Properties_x', 'CNA', 'Gene Expression_x', 'Methylation_x', 'TARGET', 'TARGET_PATHWAY', 'Sample Name', 'Whole Exome Sequencing (WES)', 'Copy Number Alterations (CNA)', 'Gene Expression_y', 'Methylation_y', 'Drug\nResponse', 'GDSC\nTissue descriptor 1', 'GDSC\nTissue\ndescriptor 2', 'Cancer Type\n(matching TCGA label)', 'Microsatellite \ninstability Status (MSI)', 'Screen Medium_y', 'Growth Properties_y']

df_encoded_categorical = pd.get_dummies(df[categorical_cols], drop_first=True)
display(df_encoded_categorical.head())

CELL_LINE_NAME_23132-87  CELL_LINE_NAME_42-MG-BA  CELL_LINE_NAME_451Lu  \
0                    False                    False                 False   
1                    False                    False                 False   
2                    False                    False                 False   
3                    False                    False                 False   
4                    False                    False                 False   

   CELL_LINE_NAME_639-V  CELL_LINE_NAME_647-V  CELL_LINE_NAME_769-P  \
0                 False                 False                 False   
1                 False                 False                 False   
2                 False                 False                 False   
3                 False                 False                 False   
4                 False                 False                 False   

   CELL_LINE_NAME_786-0  CELL_LINE_NAME_8-MG-BA  CELL_LINE_NAME_8305C  \
0                 False                   False                 False   
1                 False                   False                 False   
2                 False                   False                 False   
3                 False                   False                 False   
4                 False                   False                 False   

   CELL_LINE_NAME_8505C  ...  Cancer Type\n(matching TCGA label)_SCLC  \
0                 False  ...                                    False   
1                 False  ...                                    False   
2                 False  ...                                    False   
3                 False  ...                                    False   
4                 False  ...                                    False   

   Cancer Type\n(matching TCGA label)_SKCM  \
0                                    False   
1                                     True   
2                                    False   
3                                    False   
4                                    False   

   Cancer Type\n(matching TCGA label)_STAD  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   Cancer Type\n(matching TCGA label)_THCA  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   Cancer Type\n(matching TCGA label)_UCEC  \
0                                    False   
1                                    False   
2                                    False   
3                                    False   
4                                    False   

   Cancer Type\n(matching TCGA label)_UNABLE TO CLASSIFY  \
0                                              False       
1                                              False       
2                                              False       
3                                              False       
4                                              False       

   Microsatellite \ninstability Status (MSI)_MSS/MSI-L  Screen Medium_y_R  \
0                                               True                 True   
1                                               True                 True   
2                                               True                False   
3                                               True                False   
4                                               True                False   

   Growth Properties_y_Semi-Adherent  Growth Properties_y_Suspension  
0                              False                           False  
1                              False                           False  
2                              False                           False  
3                  

## (optional) dimensionality reduction

### Subtask:
Apply dimensionality reduction techniques like PCA or autoencoders to high-dimensional features if necessary.


**Reasoning**:
Combine the scaled numerical features and the one-hot encoded categorical features into a single DataFrame. Given the potentially large number of features after one-hot encoding (2160 columns from `df_encoded_categorical` alone), dimensionality reduction might be beneficial. I will combine the features first to assess the total dimensionality and then determine if PCA is necessary.



In [10]:
df_combined_features = pd.concat([df_scaled_numerical, df_encoded_categorical], axis=1)
print(f"Shape of the combined feature DataFrame: {df_combined_features.shape}")

Shape of the combined feature DataFrame: (166644, 2166)


**Reasoning**:
The combined feature set has 2166 columns, which is high-dimensional. Applying PCA is a suitable dimensionality reduction technique in this case to reduce the number of features while retaining most of the variance.



In [11]:
from sklearn.decomposition import PCA

# Determine the number of components. A common approach is to retain a certain percentage of variance.
# Let's aim to retain 95% of the variance.
n_components = 0.95
pca = PCA(n_components=n_components)

# Apply PCA to the combined feature DataFrame
df_pca = pca.fit_transform(df_combined_features)

# Convert the PCA results back to a DataFrame for easier handling
df_pca = pd.DataFrame(df_pca, columns=[f'PCA_{i}' for i in range(df_pca.shape[1])])

print(f"Shape of the DataFrame after PCA: {df_pca.shape}")
display(df_pca.head())

Shape of the DataFrame after PCA: (166644, 619)


PCA_0     PCA_1     PCA_2     PCA_3     PCA_4     PCA_5     PCA_6  \
0 -0.132414 -2.024178 -1.540078 -0.049719  0.191873  0.282552  0.314325   
1 -0.259444 -2.044082 -1.588382  0.201309  0.178479  0.317194  0.821190   
2 -1.128180 -2.328082 -1.490207  0.631371 -0.757776 -0.163597 -0.273552   
3 -0.110659 -2.020106 -1.656911  0.639512 -0.639597  0.054485 -0.242050   
4 -0.759971 -2.217743 -1.556301  0.666106 -0.697898 -0.057881 -0.258839   

      PCA_7     PCA_8     PCA_9  ...       PCA_609       PCA_610  \
0 -0.150184  0.012957  0.743412  ... -1.241950e-14 -1.110502e-14   
1 -0.961142  2.333903 -0.080839  ... -2.429253e-14  2.066785e-14   
2 -0.078241 -0.310717  0.467383  ...  1.138301e-01 -6.184504e-03   
3 -0.067078 -0.274854  0.613277  ...  5.487678e-02  1.062580e-01   
4 -0.072902 -0.294335  0.532888  ... -6.427011e-02 -2.070079e-01   

        PCA_611       PCA_612       PCA_613       PCA_614       PCA_615  \
0 -2.325918e-15  8.503740e-16 -7.162715e-15 -7.939013e-15  2.655723e-16   
1  2.549197e-14  3.956078e-14  3.838970e-15  2.202886e-14 -4.507939e-14   
2  1.777257e-01 -5.722606e-02  1.362491e-02  3.768531e-02 -6.291461e-02   
3 -1.275984e-02 -1.337027e-01 -6.678030e-02 -8.630510e-02  1.342935e-01   
4 -3.143296e-01  1.219717e-01 -2.199954e-01 -2.335864e-02  2.309084e-01   

        PCA_616       PCA_617       PCA_618  
0 -2.450442e-15  1.154416e-14 -5.512259e-16  
1  5.751174e-14  2.711294e-14 -4.811695e-14  
2  3.357188e-01 -2.814202e-01 -4.333066e-03  
3 -2.318729e-01  1.290547e-01 -1.759355e-02  
4 -1.833547e-01  6.339183e-02  1.432502e-02  

[5 rows x 619 columns]

## Combine features

### Subtask:
Combine the processed numerical, categorical, and drug features into a single DataFrame.


**Reasoning**:
Concatenate the scaled numerical features and the dimensionality-reduced features into a single DataFrame, then display its head and shape.



In [12]:
df_processed = pd.concat([df_scaled_numerical, df_pca], axis=1)
display(df_processed.head())
print("Shape of the processed DataFrame:", df_processed.shape)

COSMIC_ID   DRUG_ID   LN_IC50       AUC   Z_SCORE  COSMIC identifier  \
0  -1.348967 -1.364272 -1.523914  0.333408  0.390658          -1.348967   
1  -1.332118 -1.364272 -1.442914 -0.091805  0.515240          -1.332118   
2  -1.332087 -1.364272 -2.054572 -0.402301 -0.425523          -1.332087   
3  -1.332078 -1.364272 -1.518694  0.163179  0.398688          -1.332078   
4  -1.332069 -1.364272 -1.837766 -0.253566 -0.092063          -1.332069   

      PCA_0     PCA_1     PCA_2     PCA_3  ...       PCA_609       PCA_610  \
0 -0.132414 -2.024178 -1.540078 -0.049719  ... -1.241950e-14 -1.110502e-14   
1 -0.259444 -2.044082 -1.588382  0.201309  ... -2.429253e-14  2.066785e-14   
2 -1.128180 -2.328082 -1.490207  0.631371  ...  1.138301e-01 -6.184504e-03   
3 -0.110659 -2.020106 -1.656911  0.639512  ...  5.487678e-02  1.062580e-01   
4 -0.759971 -2.217743 -1.556301  0.666106  ... -6.427011e-02 -2.070079e-01   

        PCA_611       PCA_612       PCA_613       PCA_614       PCA_615  \
0 -2.325918e-15  8.503740e-16 -7.162715e-15 -7.939013e-15  2.655723e-16   
1  2.549197e-14  3.956078e-14  3.838970e-15  2.202886e-14 -4.507939e-14   
2  1.777257e-01 -5.722606e-02  1.362491e-02  3.768531e-02 -6.291461e-02   
3 -1.275984e-02 -1.337027e-01 -6.678030e-02 -8.630510e-02  1.342935e-01   
4 -3.143296e-01  1.219717e-01 -2.199954e-01 -2.335864e-02  2.309084e-01   

        PCA_616       PCA_617       PCA_618  
0 -2.450442e-15  1.154416e-14 -5.512259e-16  
1  5.751174e-14  2.711294e-14 -4.811695e-14  
2  3.357188e-01 -2.814202e-01 -4.333066e-03  
3 -2.318729e-01  1.290547e-01 -1.759355e-02  
4 -1.833547e-01  6.339183e-02  1.432502e-02  

[5 rows x 625 columns]

Shape of the processed DataFrame: (166644, 625)


**Reasoning**:
The previous code block failed because pandas was not imported. This code block will import pandas and then concatenate the scaled numerical features and the dimensionality-reduced features into a single DataFrame, display its head and shape.



In [13]:
import pandas as pd

df_processed = pd.concat([df_scaled_numerical, df_pca], axis=1)
display(df_processed.head())
print("Shape of the processed DataFrame:", df_processed.shape)

COSMIC_ID   DRUG_ID   LN_IC50       AUC   Z_SCORE  COSMIC identifier  \
0  -1.348967 -1.364272 -1.523914  0.333408  0.390658          -1.348967   
1  -1.332118 -1.364272 -1.442914 -0.091805  0.515240          -1.332118   
2  -1.332087 -1.364272 -2.054572 -0.402301 -0.425523          -1.332087   
3  -1.332078 -1.364272 -1.518694  0.163179  0.398688          -1.332078   
4  -1.332069 -1.364272 -1.837766 -0.253566 -0.092063          -1.332069   

      PCA_0     PCA_1     PCA_2     PCA_3  ...       PCA_609       PCA_610  \
0 -0.132414 -2.024178 -1.540078 -0.049719  ... -1.241950e-14 -1.110502e-14   
1 -0.259444 -2.044082 -1.588382  0.201309  ... -2.429253e-14  2.066785e-14   
2 -1.128180 -2.328082 -1.490207  0.631371  ...  1.138301e-01 -6.184504e-03   
3 -0.110659 -2.020106 -1.656911  0.639512  ...  5.487678e-02  1.062580e-01   
4 -0.759971 -2.217743 -1.556301  0.666106  ... -6.427011e-02 -2.070079e-01   

        PCA_611       PCA_612       PCA_613       PCA_614       PCA_615  \
0 -2.325918e-15  8.503740e-16 -7.162715e-15 -7.939013e-15  2.655723e-16   
1  2.549197e-14  3.956078e-14  3.838970e-15  2.202886e-14 -4.507939e-14   
2  1.777257e-01 -5.722606e-02  1.362491e-02  3.768531e-02 -6.291461e-02   
3 -1.275984e-02 -1.337027e-01 -6.678030e-02 -8.630510e-02  1.342935e-01   
4 -3.143296e-01  1.219717e-01 -2.199954e-01 -2.335864e-02  2.309084e-01   

        PCA_616       PCA_617       PCA_618  
0 -2.450442e-15  1.154416e-14 -5.512259e-16  
1  5.751174e-14  2.711294e-14 -4.811695e-14  
2  3.357188e-01 -2.814202e-01 -4.333066e-03  
3 -2.318729e-01  1.290547e-01 -1.759355e-02  
4 -1.833547e-01  6.339183e-02  1.432502e-02  

[5 rows x 625 columns]

Shape of the processed DataFrame: (166644, 625)


**Reasoning**:
The previous code block failed because `df_scaled_numerical` and `df_pca` were not defined in the current session. This code block will regenerate `df_scaled_numerical` and `df_pca` and then concatenate them into a single DataFrame, display its head and shape.



In [14]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

numerical_cols = ['COSMIC_ID', 'DRUG_ID', 'LN_IC50', 'AUC', 'Z_SCORE', 'COSMIC identifier']
scaler = StandardScaler()
df_scaled_numerical = pd.DataFrame(scaler.fit_transform(df[numerical_cols]), columns=numerical_cols)

categorical_cols = ['CELL_LINE_NAME', 'TCGA_DESC', 'DRUG_NAME', 'GDSC Tissue descriptor 1', 'GDSC Tissue descriptor 2', 'Cancer Type (matching TCGA label)', 'Microsatellite instability Status (MSI)', 'Screen Medium_x', 'Growth Properties_x', 'CNA', 'Gene Expression_x', 'Methylation_x', 'TARGET', 'TARGET_PATHWAY', 'Sample Name', 'Whole Exome Sequencing (WES)', 'Copy Number Alterations (CNA)', 'Gene Expression_y', 'Methylation_y', 'Drug\nResponse', 'GDSC\nTissue descriptor 1', 'GDSC\nTissue\ndescriptor 2', 'Cancer Type\n(matching TCGA label)', 'Microsatellite \ninstability Status (MSI)', 'Screen Medium_y', 'Growth Properties_y']
df_encoded_categorical = pd.get_dummies(df[categorical_cols], drop_first=True)

df_combined_features = pd.concat([df_scaled_numerical, df_encoded_categorical], axis=1)

n_components = 0.95
pca = PCA(n_components=n_components)
df_pca = pca.fit_transform(df_combined_features)
df_pca = pd.DataFrame(df_pca, columns=[f'PCA_{i}' for i in range(df_pca.shape[1])])

df_processed = pd.concat([df_scaled_numerical, df_pca], axis=1)
display(df_processed.head())
print("Shape of the processed DataFrame:", df_processed.shape)

COSMIC_ID   DRUG_ID   LN_IC50       AUC   Z_SCORE  COSMIC identifier  \
0  -1.348967 -1.364272 -1.523914  0.333408  0.390658          -1.348967   
1  -1.332118 -1.364272 -1.442914 -0.091805  0.515240          -1.332118   
2  -1.332087 -1.364272 -2.054572 -0.402301 -0.425523          -1.332087   
3  -1.332078 -1.364272 -1.518694  0.163179  0.398688          -1.332078   
4  -1.332069 -1.364272 -1.837766 -0.253566 -0.092063          -1.332069   

      PCA_0     PCA_1     PCA_2     PCA_3  ...       PCA_609       PCA_610  \
0 -0.132414 -2.024178 -1.540078 -0.049719  ... -1.241950e-14 -1.110502e-14   
1 -0.259444 -2.044082 -1.588382  0.201309  ... -2.429253e-14  2.066785e-14   
2 -1.128180 -2.328082 -1.490207  0.631371  ...  1.138301e-01 -6.184504e-03   
3 -0.110659 -2.020106 -1.656911  0.639512  ...  5.487678e-02  1.062580e-01   
4 -0.759971 -2.217743 -1.556301  0.666106  ... -6.427011e-02 -2.070079e-01   

        PCA_611       PCA_612       PCA_613       PCA_614       PCA_615  \
0 -2.325918e-15  8.503740e-16 -7.162715e-15 -7.939013e-15  2.655723e-16   
1  2.549197e-14  3.956078e-14  3.838970e-15  2.202886e-14 -4.507939e-14   
2  1.777257e-01 -5.722606e-02  1.362491e-02  3.768531e-02 -6.291461e-02   
3 -1.275984e-02 -1.337027e-01 -6.678030e-02 -8.630510e-02  1.342935e-01   
4 -3.143296e-01  1.219717e-01 -2.199954e-01 -2.335864e-02  2.309084e-01   

        PCA_616       PCA_617       PCA_618  
0 -2.450442e-15  1.154416e-14 -5.512259e-16  
1  5.751174e-14  2.711294e-14 -4.811695e-14  
2  3.357188e-01 -2.814202e-01 -4.333066e-03  
3 -2.318729e-01  1.290547e-01 -1.759355e-02  
4 -1.833547e-01  6.339183e-02  1.432502e-02  

[5 rows x 625 columns]

Shape of the processed DataFrame: (166644, 625)


In [15]:
from sklearn.model_selection import StratifiedShuffleSplit
merged_data = pd.read_csv("C:\\Users\\vinay\\Downloads\\merged_data.csv")

# Define the target variable (assuming 'LN_IC50' is the target)
# You might need to adjust this based on your specific modeling goal
target = merged_data['LN_IC50']
features = merged_data.drop(columns=['LN_IC50']) # Drop the target and potentially other non-feature columns

# Use the original 'TCGA_DESC' column from the 'merged' DataFrame for stratification
stratify_col = merged_data['TCGA_DESC']


# Split into training (60%) and temporary (40%) sets, stratified by 'TCGA_DESC'
# We use n_splits=1 because we only need one split for train/temp
split1 = StratifiedShuffleSplit(n_splits=1, test_size=0.4, random_state=42)

for train_index, temp_index in split1.split(features, stratify_col):
    X_train, X_temp = features.iloc[train_index], features.iloc[temp_index]
    y_train, y_temp = target.iloc[train_index], target.iloc[temp_index]
    stratify_temp = stratify_col.iloc[temp_index] # Stratify the temp split

# Split the temporary set into validation (50% of temp, 20% of total) and test (50% of temp, 20% of total)
split2 = StratifiedShuffleSplit(n_splits=1, test_size=0.5, random_state=42)

for val_index, test_index in split2.split(X_temp, stratify_temp):
    X_val, X_test = X_temp.iloc[val_index], X_temp.iloc[test_index]
    y_val, y_test = y_temp.iloc[val_index], y_temp.iloc[test_index]

# Print the shapes of the resulting sets
print("Shape of X_train:", X_train.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of X_val:", X_val.shape)
print("Shape of y_val:", y_val.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (99986, 31)
Shape of y_train: (99986,)
Shape of X_val: (33329, 31)
Shape of y_val: (33329,)
Shape of X_test: (33329, 31)
Shape of y_test: (33329,)


In [18]:
%pip install matplotlib
%pip install seaborn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 24.0 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [20]:
import numpy as np
import pandas as pd
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import ElasticNet, LinearRegression
from xgboost import XGBRegressor
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.preprocessing import StandardScaler
import warnings
warnings.filterwarnings('ignore')

# ============================================
# 1. LOAD AND SPLIT DATA
# ============================================

# Load data
merged_data = pd.read_csv("C:\\Users\\vinay\\OneDrive\\Desktop\\hackathon\\autopharma\\dataset\\merged_data.csv")

# Define target and features
target = merged_data['LN_IC50']
features = merged_data.drop(columns=['LN_IC50'])
stratify_col = merged_data['TCGA_DESC']

print("Original data shape:", merged_data.shape)
print("Features shape:", features.shape)
print("Target shape:", target.shape)
print("\nTCGA_DESC distribution:")
print(stratify_col.value_counts())

# Split into training (60%) and temporary (40%) sets
split1 = StratifiedShuffleSplit(n_splits=1, test_size=0.4, random_state=42)

for train_index, temp_index in split1.split(features, stratify_col):
    X_train, X_temp = features.iloc[train_index], features.iloc[temp_index]
    y_train, y_temp = target.iloc[train_index], target.iloc[temp_index]
    stratify_temp = stratify_col.iloc[temp_index]

# Split temporary into validation (20%) and test (20%)
split2 = StratifiedShuffleSplit(n_splits=1, test_size=0.5, random_state=42)

for val_index, test_index in split2.split(X_temp, stratify_temp):
    X_val, X_test = X_temp.iloc[val_index], X_temp.iloc[test_index]
    y_val, y_test = y_temp.iloc[val_index], y_temp.iloc[test_index]

print("\n" + "="*50)
print("DATA SPLIT SUMMARY")
print("="*50)
print(f"Train: {X_train.shape[0]} samples ({X_train.shape[0]/len(features)*100:.1f}%)")
print(f"Val:   {X_val.shape[0]} samples ({X_val.shape[0]/len(features)*100:.1f}%)")
print(f"Test:  {X_test.shape[0]} samples ({X_test.shape[0]/len(features)*100:.1f}%)")

# ============================================
# 2. PREPROCESSING
# ============================================

# Identify numeric columns (exclude categorical like TCGA_DESC, DRUG_NAME, etc.)
categorical_cols = X_train.select_dtypes(include=['object']).columns.tolist()
numeric_cols = X_train.select_dtypes(include=[np.number]).columns.tolist()

print(f"\nNumeric features: {len(numeric_cols)}")
print(f"Categorical features: {len(categorical_cols)}")

# Handle categorical columns (one-hot encoding or dropping)
if categorical_cols:
    print(f"\nCategorical columns found: {categorical_cols[:5]}...")
    # Option 1: Drop categorical columns for simple model
    X_train_num = X_train[numeric_cols]
    X_val_num = X_val[numeric_cols]
    X_test_num = X_test[numeric_cols]
    
    # Option 2: One-hot encode (uncomment if you want to use categorical features)
    # X_train_num = pd.get_dummies(X_train, columns=categorical_cols, drop_first=True)
    # X_val_num = pd.get_dummies(X_val, columns=categorical_cols, drop_first=True)
    # X_test_num = pd.get_dummies(X_test, columns=categorical_cols, drop_first=True)
    # # Align columns
    # X_val_num = X_val_num.reindex(columns=X_train_num.columns, fill_value=0)
    # X_test_num = X_test_num.reindex(columns=X_train_num.columns, fill_value=0)
else:
    X_train_num = X_train
    X_val_num = X_val
    X_test_num = X_test

# Standardize features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_num)
X_val_scaled = scaler.transform(X_val_num)
X_test_scaled = scaler.transform(X_test_num)

print(f"\nFinal feature dimensions after preprocessing: {X_train_scaled.shape[1]}")

# ============================================
# 3. MODEL TRAINING & EVALUATION
# ============================================

from scipy.stats import pearsonr

def evaluate_model(model, X, y, set_name=""):
    """Evaluate model and return metrics"""
    y_pred = model.predict(X)
    mse = mean_squared_error(y, y_pred)
    rmse = np.sqrt(mse)
    mae = mean_absolute_error(y, y_pred)
    r2 = r2_score(y, y_pred)
    pearson_corr, pearson_pval = pearsonr(y, y_pred)
    
    print(f"\n{set_name} Metrics:")
    print(f"  RMSE:    {rmse:.4f}")
    print(f"  MAE:     {mae:.4f}")
    print(f"  R²:      {r2:.4f}")
    print(f"  Pearson: {pearson_corr:.4f} (p={pearson_pval:.2e})")
    
    return {'RMSE': rmse, 'MAE': mae, 'R2': r2, 'Pearson': pearson_corr, 'predictions': y_pred}

print("\n" + "="*50)
print("MODEL TRAINING")
print("="*50)

results = {}

# ============================================
# BASELINE: Linear Regression
# ============================================
print("\n[1/4] Training Linear Regression (Baseline)...")
lr_model = LinearRegression()
lr_model.fit(X_train_scaled, y_train)

results['Linear Regression'] = {
    'model': lr_model,
    'train': evaluate_model(lr_model, X_train_scaled, y_train, "Train"),
    'val': evaluate_model(lr_model, X_val_scaled, y_val, "Validation"),
    'test': evaluate_model(lr_model, X_test_scaled, y_test, "Test")
}

# ============================================
# ElasticNet
# ============================================
print("\n[2/4] Training ElasticNet...")
elastic_model = ElasticNet(alpha=0.1, l1_ratio=0.5, max_iter=10000, random_state=42)
elastic_model.fit(X_train_scaled, y_train)

results['ElasticNet'] = {
    'model': elastic_model,
    'train': evaluate_model(elastic_model, X_train_scaled, y_train, "Train"),
    'val': evaluate_model(elastic_model, X_val_scaled, y_val, "Validation"),
    'test': evaluate_model(elastic_model, X_test_scaled, y_test, "Test")
}

# ============================================
# Random Forest
# ============================================
print("\n[3/4] Training Random Forest...")
rf_model = RandomForestRegressor(
    n_estimators=100,
    max_depth=15,
    min_samples_split=10,
    min_samples_leaf=5,
    random_state=42,
    n_jobs=-1
)
rf_model.fit(X_train_scaled, y_train)

results['Random Forest'] = {
    'model': rf_model,
    'train': evaluate_model(rf_model, X_train_scaled, y_train, "Train"),
    'val': evaluate_model(rf_model, X_val_scaled, y_val, "Validation"),
    'test': evaluate_model(rf_model, X_test_scaled, y_test, "Test")
}

# ============================================
# XGBoost
# ============================================
print("\n[4/4] Training XGBoost...")
xgb_model = XGBRegressor(
    n_estimators=100,
    max_depth=6,
    learning_rate=0.1,
    subsample=0.8,
    colsample_bytree=0.8,
    random_state=42,
    n_jobs=-1
)
xgb_model.fit(X_train_scaled, y_train)

results['XGBoost'] = {
    'model': xgb_model,
    'train': evaluate_model(xgb_model, X_train_scaled, y_train, "Train"),
    'val': evaluate_model(xgb_model, X_val_scaled, y_val, "Validation"),
    'test': evaluate_model(xgb_model, X_test_scaled, y_test, "Test")
}

# ============================================
# HYPERPARAMETER TUNING (using validation set)
# ============================================
print("\n" + "="*50)
print("HYPERPARAMETER TUNING")
print("="*50)

from sklearn.model_selection import RandomizedSearchCV

# Tune Random Forest
print("\nTuning Random Forest...")
rf_param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [10, 15, 20, None],
    'min_samples_split': [5, 10, 20],
    'min_samples_leaf': [2, 5, 10]
}

# Use validation set for tuning by creating custom CV split
from sklearn.model_selection import PredefinedSplit
# Combine train and val for tuning
X_train_val = np.vstack([X_train_scaled, X_val_scaled])
y_train_val = np.concatenate([y_train, y_val])
# Create split indices: -1 for train, 0 for val
split_index = [-1] * len(X_train_scaled) + [0] * len(X_val_scaled)
ps = PredefinedSplit(test_fold=split_index)

rf_random = RandomizedSearchCV(
    RandomForestRegressor(random_state=42, n_jobs=-1),
    param_distributions=rf_param_grid,
    n_iter=10,
    cv=ps,
    scoring='neg_root_mean_squared_error',
    random_state=42,
    n_jobs=-1
)
rf_random.fit(X_train_val, y_train_val)

print(f"Best RF params: {rf_random.best_params_}")
print(f"Best RF validation RMSE: {-rf_random.best_score_:.4f}")

results['Random Forest (Tuned)'] = {
    'model': rf_random.best_estimator_,
    'val': evaluate_model(rf_random.best_estimator_, X_val_scaled, y_val, "Validation"),
    'test': evaluate_model(rf_random.best_estimator_, X_test_scaled, y_test, "Test")
}

# Tune XGBoost
print("\nTuning XGBoost...")
xgb_param_grid = {
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 6, 9],
    'learning_rate': [0.01, 0.1, 0.3],
    'subsample': [0.6, 0.8, 1.0],
    'colsample_bytree': [0.6, 0.8, 1.0]
}

xgb_random = RandomizedSearchCV(
    XGBRegressor(random_state=42, n_jobs=-1),
    param_distributions=xgb_param_grid,
    n_iter=10,
    cv=ps,
    scoring='neg_root_mean_squared_error',
    random_state=42,
    n_jobs=-1
)
xgb_random.fit(X_train_val, y_train_val)

print(f"Best XGB params: {xgb_random.best_params_}")
print(f"Best XGB validation RMSE: {-xgb_random.best_score_:.4f}")

results['XGBoost (Tuned)'] = {
    'model': xgb_random.best_estimator_,
    'val': evaluate_model(xgb_random.best_estimator_, X_val_scaled, y_val, "Validation"),
    'test': evaluate_model(xgb_random.best_estimator_, X_test_scaled, y_test, "Test")
}

# ============================================
# 4. RESULTS SUMMARY
# ============================================
print("\n" + "="*50)
print("FINAL RESULTS SUMMARY")
print("="*50)

summary_df = pd.DataFrame({
    'Model': [],
    'Val_RMSE': [],
    'Val_Pearson': [],
    'Val_R2': [],
    'Test_RMSE': [],
    'Test_Pearson': [],
    'Test_R2': []
})

for model_name, result in results.items():
    summary_df = pd.concat([summary_df, pd.DataFrame({
        'Model': [model_name],
        'Val_RMSE': [result['val']['RMSE']],
        'Val_Pearson': [result['val']['Pearson']],
        'Val_R2': [result['val']['R2']],
        'Test_RMSE': [result['test']['RMSE']],
        'Test_Pearson': [result['test']['Pearson']],
        'Test_R2': [result['test']['R2']]
    })], ignore_index=True)

print("\n", summary_df.to_string(index=False))

# Find best model based on validation RMSE
best_model_name = summary_df.loc[summary_df['Val_RMSE'].idxmin(), 'Model']
print(f"\n🏆 Best Model (by Validation RMSE): {best_model_name}")
print(f"   Test RMSE:    {results[best_model_name]['test']['RMSE']:.4f}")
print(f"   Test Pearson: {results[best_model_name]['test']['Pearson']:.4f}")
print(f"   Test R²:      {results[best_model_name]['test']['R2']:.4f}")

# ============================================
# 5. FEATURE IMPORTANCE (for tree-based models)
# ============================================
if best_model_name in ['Random Forest', 'XGBoost']:
    print("\n" + "="*50)
    print("TOP 10 FEATURE IMPORTANCES")
    print("="*50)
    
    importances = results[best_model_name]['model'].feature_importances_
    feature_importance_df = pd.DataFrame({
        'Feature': numeric_cols,
        'Importance': importances
    }).sort_values('Importance', ascending=False).head(10)
    
    print("\n", feature_importance_df.to_string(index=False))

print("\n" + "="*50)
print("PIPELINE COMPLETE!")
print("="*50)

Original data shape: (166644, 32)
Features shape: (166644, 31)
Target shape: (166644,)

TCGA_DESC distribution:
TCGA_DESC
LUAD            13441
SCLC            12041
BRCA            11289
SKCM            10754
COREAD          10659
ESCA             8008
HNSC             7811
GBM              7475
OV               7306
DLBC             7062
KIRC             6810
NB               6721
ALL              5947
PAAD             5944
LAML             5205
STAD             4289
MM               3834
BLCA             3719
LIHC             3384
LUSC             3337
THCA             3328
CESC             3324
LGG              3184
UNCLASSIFIED     2863
LCML             2272
UCEC             2174
PRAD             1451
MESO             1371
MB                924
CLL               475
ACC               242
Name: count, dtype: int64

DATA SPLIT SUMMARY
Train: 99986 samples (60.0%)
Val:   33329 samples (20.0%)
Test:  33329 samples (20.0%)

Numeric features: 5
Categorical features: 26

Categorical colu

In [21]:
import pickle
import os

# ============================================
# EXPORT TRAINED MODELS
# ============================================
print("\n" + "="*50)
print("EXPORTING MODELS")
print("="*50)

# Define save directory
save_dir = "C:\\Users\\vinay\\OneDrive\\Desktop\\hackathon\\autopharma\\trained_models"

# Create directory if it doesn't exist
os.makedirs(save_dir, exist_ok=True)
print(f"\nSave directory: {save_dir}")

# Export the best model
best_model = results[best_model_name]['model']
best_model_path = os.path.join(save_dir, f'best_model_{best_model_name.replace(" ", "_")}.pkl')
with open(best_model_path, 'wb') as f:
    pickle.dump(best_model, f)
print(f"✅ Best model saved: {best_model_path}")

# Export the scaler (important for prediction!)
scaler_path = os.path.join(save_dir, 'scaler.pkl')
with open(scaler_path, 'wb') as f:
    pickle.dump(scaler, f)
print(f"✅ Scaler saved: {scaler_path}")

# Optional: Export all models
for model_name, result in results.items():
    filename = f'model_{model_name.replace(" ", "_").replace("(", "").replace(")", "")}.pkl'
    filepath = os.path.join(save_dir, filename)
    with open(filepath, 'wb') as f:
        pickle.dump(result['model'], f)
    print(f"✅ {model_name} saved: {filepath}")

# Export model metadata
metadata = {
    'best_model_name': best_model_name,
    'feature_columns': numeric_cols,
    'results_summary': summary_df.to_dict(),
    'training_date': pd.Timestamp.now().strftime('%Y-%m-%d %H:%M:%S')
}
metadata_path = os.path.join(save_dir, 'model_metadata.pkl')
with open(metadata_path, 'wb') as f:
    pickle.dump(metadata, f)
print(f"✅ Metadata saved: {metadata_path}")

print("\n" + "="*50)
print(f"ALL MODELS SAVED TO: {save_dir}")
print("="*50)


EXPORTING MODELS

Save directory: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models
✅ Best model saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\best_model_XGBoost_(Tuned).pkl
✅ Scaler saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\scaler.pkl
✅ Linear Regression saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\model_Linear_Regression.pkl
✅ ElasticNet saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\model_ElasticNet.pkl
✅ Random Forest saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\model_Random_Forest.pkl
✅ XGBoost saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\model_XGBoost.pkl
✅ Random Forest (Tuned) saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\model_Random_Forest_Tuned.pkl
✅ XGBoost (Tuned) saved: C:\Users\vinay\OneDrive\Desktop\hackathon\autopharma\trained_models\model_XGBoo